metadata_store_scripts/V2025.07.22.0__add_generate_masking_parameters_procedure.sql

```diff
@@ -0,0 +1,391 @@
+ *   - Applies conditional masking based on key column values
+ *   - Uses alias-specific algorithms and date formats
+ *   - Filters rules to match the specified condition alias
+ *
+ * OUTPUT PARAMETERS:
+ * - FieldAlgorithmAssignments: JSON mapping of encoded column names to masking algorithms
+ * - ColumnsToMask: JSON array of column names requiring masking
+ * - DataFactoryTypeMapping: ADF schema definition for API response parsing
+ * - NumberOfBatches: Calculated batch count for optimal processing
+ * - TrimLengths: JSON array of column max lengths for output trimming
+ * - DateFormatAssignments: JSON mapping of columns to their date format strings
+ * - ColumnsToCastAsStrings: JSON array of columns requiring string casting
+ * - ColumnsToCastBackTo*: JSON arrays of columns to cast back to specific ADF types
+ *
+ * DEPENDENCIES:
+ * - discovered_ruleset table: Contains masking rules and algorithm assignments
+ * - adf_type_mapping table: Maps database types to ADF types
+ * - Supports conditional masking via JSON structures in assigned_algorithm column
+ *
+ * FALLBACK BEHAVIOR:
+ * Always returns exactly one record. If no masking rules exist, returns default/empty
+ * values to prevent ADF lookup activity failures.
+ *
+ * NOTES:
+ * - Empty JSON arrays are converted to [""] for ADF UI compatibility
+ * - Uses hex-encoded column names to avoid ADF pipeline issues with special characters
+ * - Supports treat_as_string flag for type casting feature
+ */
+CREATE OR ALTER PROCEDURE generate_masking_parameters
+    @source_database NVARCHAR(128),
+    @source_schema NVARCHAR(128),
+    @source_table NVARCHAR(128),
+    @dataset NVARCHAR(128),
+    @column_width_estimate INT = 1000,
+    @filter_key NVARCHAR(128) = ''  -- Optional filter key for conditional masking
+AS
+BEGIN
+    SET NOCOUNT ON;
+
+    -- base_ruleset_filter - Filter to target table, add ADF type mappings
+    WITH base_ruleset_filter AS (
+        SELECT
+            r.*,
+            t.adf_type
+        FROM discovered_ruleset AS r
+        INNER JOIN adf_type_mapping AS t
+            ON
+                r.identified_column_type = t.dataset_type
+                AND r.dataset = t.dataset
+        WHERE
+            r.dataset = @dataset
+            AND r.specified_database = @source_database
+            AND r.specified_schema = @source_schema
+            AND r.identified_table = @source_table
+            AND r.assigned_algorithm IS NOT NULL
+            AND r.assigned_algorithm <> ''
+    ),
+
+    -- ruleset_computed - Add encoded column names and extract JSON metadata
+    ruleset_computed AS (
+        SELECT
+            r.*,
+            CONCAT(
+                'x',
+                CONVERT(
+                    VARCHAR(MAX),
+                    CONVERT(VARBINARY, r.identified_column),
+                    2
+                )
+            ) AS encoded_column_name,
+            JSON_VALUE(r.algorithm_metadata, '$.date_format') AS [date_format],
```

Comment:
question: why is it that we need a `CASE` for `r.assigned_algorithm` is JSON, but not for `r.algorithm_metadata`?